### PR TITLE
GH-1185: ensure that only one save is performed at the time

### PR DESCRIPTION
**Pull request by akosyakov**
@lmcbout, @hexa00 please try this change whether it fixes https://github.com/theia-ide/theia/issues/1185

Follow-ups discussed on Gitter:
- improve `FileResource` that it does not reread a content if a file state has not been changed
- improve the performance of sending big messages over ws, look into using https://github.com/developit/greenlet

### CODE REVIEW

**Comment by lmcbout** [comment]
@akosyakov I just tested with the new code. It is better but not perfect. I was typing slowly and it occurs again
Typing: "jacques" :), I guest it doesn't like my name :), at the end, the letter "C" is missing

peek 2018-02-01 13-27typuing testnew1188

File: packages/monaco/src/browser/monaco-editor-model.ts [hunk 1]
```diff
@@ -173,34 +173,59 @@ export class MonacoEditorModel implements ITextEditorModel, TextEditorDocument {
...
-                const content = this.model.getValue();
-                await this.resource.saveContents(content);
-                this.setDirty(false);
+    protected saved: Promise<void> = Promise.resolve();
+    protected saveCancellationTokenSource = new CancellationTokenSource();
+    protected scheduleSave(reason: TextDocumentSaveReason): Promise<void> {
+        this.saveCancellationTokenSource.cancel();
+        this.saveCancellationTokenSource = new CancellationTokenSource();
+        const token = this.saveCancellationTokenSource.token;
+        return this.saved = this.saved.then(() => this.doSave(reason, token));
```
Comment (akosyakov) [inline packages/monaco/src/browser/monaco-editor-model.ts]: from gitter by @hexa00:

> reading this it seems that a scheduleSave will cancel the previous one if it is running... ? could it make it so that continously typing autosave will not be called.. I'm not sure if it shouln't be the other way around... finish the one that was there and try to schedule one after that?
it may be fine too the moment one stops typing for 500ms it will save...

File: packages/core/src/common/cancellation.ts [hunk 2]
```diff
@@ -21,7 +21,8 @@ export interface CancellationToken {
 }
 
 const shortcutEvent: Event<void> = Object.freeze(function (callback: any, contex?: any): any {
```
Comment (szarnekow) [inline packages/core/src/common/cancellation.ts]: contex vs context - there is a typo in the parameter name

Comment (akosyakov) [reply packages/core/src/common/cancellation.ts]: that's the thing :) I was struggling to understand why a variable is unknown :)

**Comment by lmcbout** [comment]
Tested with the latest version this morning. Some improvement, but I can still lose letters. After typing "QW" with a delay between the letter, It saved it and after a short time, the editor became dirty again without doing anything and it swallowed the last letter.
I used version:
commit **fe8af48**46d73212c08056cf1a7de0a6ebd0055e7 (HEAD -> ak/one_save_at_the_time, upstream/ak/one_save_at_the_time)
Author: Anton Kosiakov <anton.kosyakov@typefox.io>
Date:   Thu Feb 1 18:29:41 2018 +0100

    GH-1185: ensure that only one save is performed at the time


Video:
peek 2018-02-01 13-27typing testnew1188_feb2

**Comment by akosyakov** [comment]
@lmcbout please try again

The issue was that fs notifications of old save operations come delayed after new operations were applied and rollback them. Now file resources ignore changes if they are outdated based on a file state.

**Comment by kittaakos** [comment]
There are two problems we have to tackle in this PR:
 - we cannot rely on the [timestamps](https://nodejs.org/en/docs/guides/working-with-different-filesystems/#timestamp-resolution)
 - we can still fall back to the `stat.size` but clearly, file with a content `abc` will have the same size as a file with `xyz` content.

**Comment by lmcbout** [comment]
still don't work
using last "commit c316e02bac6f7fa345b260b313c5997c87e97441 (HEAD -> ak/one_save_at_the_time, upstream/ak/one_save_at_the_time) "
loosing the "R" after a save while typing "qwerty"
peek 2018-02-01 13-27typing testnew1188_feb2_pase2

**Comment by kittaakos** [comment]
I am getting closer... if one puts a breakpoint to the backend's FS#setContent and type something in the editor, requests might arrive in out-of-order. I get the "override" dialog multiple times (that is another problem, I guess).

**Comment by kittaakos** [comment]
And we should not show the file name with the message in the dialog title but as the content. Also, let's show the FS path instead of the URI.
screen shot 2018-02-02 at 16 40 03

**Comment by kittaakos** [comment]
> I get the "override" dialog multiple times (that is another problem, I guess).

I was checking it on the `master`, sorry.

**Comment by akosyakov** [comment]
> And we should not show the file name with the message in the dialog title but as the content. 

We do it after VS code.

**Comment by kittaakos** [comment]
> We do it after VS code.

It still looks alien. Anyway, that is a different story.

File: packages/filesystem/src/browser/file-resource.ts [hunk 3]
```diff
@@ -35,29 +37,55 @@ export class FileResource implements Resource {
...
+        this.state = { stat, content };
+    }
+    protected async doSaveContents(content: string, options?: { encoding?: string }): Promise<FileStat> {
         const uri = this.uri.toString();
-        if (await this.fileSystem.exists(uri)) {
-            this.stat = this.stat || await this.fileSystem.getFileStat(this.uri.toString());
-            this.stat = await this.fileSystem.setContent(this.stat, content, options);
-        } else {
-            await this.fileSystem.createFile(uri, { content });
+        if (!await this.fileSystem.exists(uri)) {
```
Comment (kittaakos) [inline packages/filesystem/src/browser/file-resource.ts]: I would instead always try to `getFileStat` and catch the error when the file does not exist. If we have the error, then let's create the new file. So we could save one `exist` check each time.

This check is not atomic anyway, so even if it returns with `true` for the existence check, someone else could delete the file before we write it. What do you think?

Comment (akosyakov) [reply packages/filesystem/src/browser/file-resource.ts]: I know, but when we will always log an error in the backend, it is not nice

Comment (kittaakos) [reply packages/filesystem/src/browser/file-resource.ts]: And I have another tiny tiny feedback to this class in general; I would save a uri.toSrting() as a variable. Parsing the URI should not be expensive, but storing it is for sure cheaper.

Comment (kittaakos) [reply packages/filesystem/src/browser/file-resource.ts]: > I know, but then we will always log an error in the backend, it is not nice

I see your point. What about changing the API then? :)
`getFileStat(uri: string): Promise<FileStat>` => `getFileStat(uri: string): Promise<FileStat | undefined>`

Comment (kittaakos) [reply packages/filesystem/src/browser/file-resource.ts]: OK, let's put these aside until we fix the original issue. Sorry.

Comment (akosyakov) [reply packages/filesystem/src/browser/file-resource.ts]: let's file an issue for it, also for `setContents` maybe it can create a file if does not exist

**Comment by akosyakov** [comment]
@lmcbout please pull and try again

File: packages/monaco/src/browser/monaco-editor-model.ts [hunk 4]
```diff
@@ -158,12 +158,25 @@ export class MonacoEditorModel implements ITextEditorModel, TextEditorDocument {
...
 
-        this.model.applyEdits([this.p2m.asTextEdit({
+        this.applyEdits([this.p2m.asTextEdit({
             range: this.m2p.asRange(this.model.getFullModelRange()),
             newText
         }) as monaco.editor.IIdentifiedSingleEditOperation]);
     }
+
+    protected ignoreEdits = false;
+    protected applyEdits(operations: monaco.editor.IIdentifiedSingleEditOperation[]): monaco.editor.IIdentifiedSingleEditOperation[] {
```
Comment (hexa00) [inline packages/monaco/src/browser/monaco-editor-model.ts]: I don't understand that one , is it that applyEdits is async somehow?

**Comment by hexa00** [comment]
@akosyakov it seems to still re-read the file.

If i open a 3MB file and type "a" in it and wait this is what I see in ws debug:

autosave

Seems like the file is still fetched...

**Comment by lmcbout** [comment]
Same issue, does not work
git log : commit **e92040eb**8ae3d7abe7f76f843c6b62e93f9c1c1b (HEAD -> ak/one_save_at_the_time, upstream/ak/one_save_at_the_time)
Video: The letter disappear 
peek 2018-02-01 13-27typing testnew1188_feb2_pase3

File: packages/filesystem/src/browser/file-resource.ts [hunk 5]
```diff
@@ -35,30 +39,55 @@ export class FileResource implements Resource {
...
+    protected async isOutdated(): Promise<boolean> {
+        if (!this.state.stat) {
+            return true;
         }
+        const stat = await this.fileSystem.getFileStat(this.uriString);
+        /*
+         * if timestamps equal we cannot assume that content has not been changed because of
+         * https://nodejs.org/en/docs/guides/working-with-different-filesystems/#timestamp-resolution
+         */
+        return stat.lastModification >= this.state.stat.lastModification;
```
Comment (hexa00) [inline packages/filesystem/src/browser/file-resource.ts]: See https://github.com/theia-ide/theia/pull/1188#issuecomment-362659715 if we are the only ones to save then the timestamp will always be equal... so this isOutdated is always true.

Should we check the contents with a sha1 for example instead or a md5 ?

File: packages/monaco/src/browser/monaco-editor-model.ts (hunk 4 above)
Comment (akosyakov) [reply packages/monaco/src/browser/monaco-editor-model.ts]: Yes, it is performed either by a fs change or during a save operation and consequently should not make document dirty and trigger a new save. (the document is already saved or will be saved).

File: packages/filesystem/src/browser/file-resource.ts (hunk 5 above)
Comment (akosyakov) [reply packages/filesystem/src/browser/file-resource.ts]: It is turned out that [my assumptions](https://github.com/theia-ide/theia/pull/1188#issuecomment-362595110) were wrong and this change does not fix an issue with the editor. I've decided to leave it since in some cases it can help. Adding hashes probably can improve even more, but again when hashes equal we cannot assume equality of content and it will be a case if one call read immediately after save.

I will look for the root cause of the original issue next week, but I don't think this code related.

Comment (hexa00) [reply packages/filesystem/src/browser/file-resource.ts]: If you use hash you can know that equally is the same content that's the point of using a hash no ?

Comment (akosyakov) [reply packages/filesystem/src/browser/file-resource.ts]: Equality of hashes does not guarantee equality of content. The point is that when hashes are not equal you know for sure that content also, and can spare yourself from the content comparison.

Comment (hexa00) [reply packages/filesystem/src/browser/file-resource.ts]: @akosyakov that's not true. Doing a SHA-256 on a file will guarantee equality of content.  Collisions can happen in theory but are very very unlikely in our use case.

File: packages/monaco/src/browser/monaco-editor-model.ts (hunk 4 above)
Comment (hexa00) [reply packages/monaco/src/browser/monaco-editor-model.ts]: Ha yes... I wonder if it would be easier to queue the save & sync methods then adding 3 mutexs  :(

Comment (akosyakov) [reply packages/monaco/src/browser/monaco-editor-model.ts]: I like an idea of having a queue of cancellable sync and save operations, then there should be no chance to have a race between them. I will give it a try.

What do you mean instead of 3 mutexs? I cannot replace 2 cancellation tokens by one since a sync operation can be canceled by fs changes as well as by document changes and a save operation can be canceled only by document changes.

File: packages/filesystem/src/browser/file-resource.ts (hunk 5 above)
Comment (akosyakov) [reply packages/filesystem/src/browser/file-resource.ts]: ok, we can have a look at it separately

File: packages/monaco/src/browser/monaco-editor-model.ts (hunk 4 above)
Comment (hexa00) [reply packages/monaco/src/browser/monaco-editor-model.ts]: Yes I did not mean replace it with one... I just meant the current structure with 2 cancellation tokens and one mutex is hard to follow. 

Seems like we need to serialize this better like with the queue idea this way we can treat theses operations as transactions ie. only remove them from the queue once all async/await calls are done for that operation and ensure that only one operation executes at a time.

Seems like it could be a useful abstraction in general too :)

**Comment by lmcbout** [comment]
Hi @akosyakov 
Just tested the latest patch with commit: 103d4b71a5c916a6ab7af36be73894b580d8fa10 
:(
It didn't work. I waited a bit to type the second letter "W" and waited, and the letter disappear after a while.
Video:
peek 2018-02-06 09-50typingnew1188_commit103d

**Comment by lmcbout** [comment]
A screen snap snapshot of the browser with the file when I ran the test. Isee the last method you created: "export class PromiseQueue"
image

**Comment by akosyakov** [comment]
@lmcbout could you add this logging at the beginning of `markAsDirty`:
```
console.log('markAsDirty, ignoreEdits: ' + this.ignoreEdits + ', version: ' + this.model.getVersionId());
```

and comment the log output here, just a part when you type `qw`, clean console before typing

**Comment by lmcbout** [comment]
No luck :(
second time I type the "W", and waited, it went away .
commit: e20abf3bc539673bdef916b308666f227f849254
peek 2018-02-06 09-50typingnew1188_commit_e20abf3

**Comment by akosyakov** [comment]
Could someone else try to reproduce the issue with this changes and look for the root cause? I cannot even reproduce it and tired of guessing.

**Comment by kittaakos** [comment]
> Could someone else try to reproduce the issue with this changes and look for the root cause? I cannot even reproduce it and tired of guessing.

I volunteer.

**Comment by akosyakov** [comment]
@lmcbout if you can debug and analyze it to find the root cause it would be great since it fails for you always

**Comment by marcdumais-work** [comment]
@akosyakov I have reproduced with the latest version of the PR (commit e20abf3bc539673bdef916b308666f227f849254), adding the console log as requested above. You can see it here: 

peek 2018-02-06 13-17

From the console (also visible in video): 
```
13:16:56.993 logger.ts:33 markAsDirty, ignoreEdits: false, version: 22
13:16:59.417 logger.ts:33 markAsDirty, ignoreEdits: false, version: 23
13:17:03.177 logger.ts:33 markAsDirty, ignoreEdits: true, version: 24
```

**Comment by kittaakos** [comment]
@marcdumais-work, where can I get the "test" workspace you guys were using? Or is there a script to generate it? Thanks!

**Comment by kittaakos** [comment]
> where can I get the "test" workspace you guys were using?

https://github.com/theia-ide/theia/issues/1185#issuecomment-362294258

**Comment by marcdumais-work** [comment]
@kittaakos I used my Theia directory as workspace, with this file in it (run gzip -d or so on it first)

[file3megs.txt.gz](https://github.com/theia-ide/theia/files/1700330/file3megs.txt.gz)

**Comment by kittaakos** [comment]
Confirmed:

screencast 2018-02-06 19-34-26

**Comment by akosyakov** [comment]
thank you for the file, turned out that the autosave was canceled too late

@lmcbout please pull and try again

**Comment by akosyakov** [comment]
please review

**Comment by akosyakov** [comment]
@lmcbout @hexa00 is it good now?

**Comment by lmcbout** [comment]
@akosyakov I don't lose  any char anymore, but one thing, after I stopped typing, I receive three identical message with the whole content of the file at the end. When the file is big, it make the browser very slow. With network latency for other people, it will be slow if we download three times the same content.
Look at the last three big message, they are identical, I verified and each one of those message have the same size, i.e. 3,000,013 bytes, 

screenshotbigfilesamemessages

**Comment by akosyakov** [comment]
let's solve one problem at the time if it fixes https://github.com/theia-ide/theia/issues/1185 I would like to get it merged and then focus on remaining issues depending on their priority

**Comment by akosyakov** [comment]
@hexa00 please review when you have time

File: packages/monaco/src/browser/monaco-editor-model.ts [hunk 6]
```diff
@@ -158,49 +176,86 @@ export class MonacoEditorModel implements ITextEditorModel, TextEditorDocument {
...
             this.onWillSaveModelEmitter.fire({
                 model, reason,
                 waitUntil: thenable =>
                     thenable.then(operations => {
-                        model.applyEdits(operations);
+                        if (token.isCancellationRequested) {
+                            resolve();
+                        }
+                        const invertedOperations = this.applyEdits(operations);
+                        token.onCancellationRequested(() => this.applyEdits(invertedOperations));
```
Comment (hexa00) [inline packages/monaco/src/browser/monaco-editor-model.ts]: Could you have the following sequence:

  -  await this.fireWillSaveModel(reason, token);
  -  await this.resource.saveContents(content);
  -  cancelSave()
  -  token.onCancellationRequested(() => this.applyEdits(invertedOperations)); executes
  - if (token.isCancellationRequested) return true

And this way saveContents will have saved the contents with the result from `this.applyEdits(operations);`

but what will be shown to the user is `this.applyEdits(invertedOperations));`

**Comment by hexa00** [comment]
I think we should remove the second commit from this PR too and tackle it separately.

**Comment by akosyakov** [comment]
> I think we should remove the second commit from this PR too and tackle it separately.

I will remove `isOutdated` check it seems to be useless indeed, but other changes in the second commit fixes https://github.com/theia-ide/theia/issues/1194

**Comment by hexa00** [comment]
@akosyakov OK that could be another PR too.. but it's OK if it desn't touch the same code at all I guess..

File: packages/monaco/src/browser/monaco-editor-model.ts (hunk 6 above)
Comment (akosyakov) [reply packages/monaco/src/browser/monaco-editor-model.ts]: It's a bit hard to reason about it since there are no clients. Thinking about it probably we should never rollback changes did in `fireWillSaveModel `. Otherwise, a user will see appearing and disappearing of changes, as well as can lose some changes, like if he types between `fireWillSaveModel` and `saveContents`.

I can comment out `fireWillSaveModel` and we handle it separately?

**Comment by lmcbout** [comment]
with "auto save = off"  receive twice the update

image

**Comment by akosyakov** [comment]
> with "auto save = off" receive twice the update

@lmcbout is it related to https://github.com/theia-ide/theia/issues/1185 or https://github.com/theia-ide/theia/issues/1225? This PR is about https://github.com/theia-ide/theia/issues/1185

**Comment by lmcbout** [comment]
@akosyakov without any auto-save, then it is related to this issue (# 1185), we should only save once.

**Comment by akosyakov** [comment]
No, it is not related to #1185. This PR is about having consistent saving and I see only one `setContent` in your snapshot, that is what save does. After that fs send notifications about changes of different kinds, on your os it sends 2 for some reasons. We decided to handle it in #1225.

**Comment by hexa00** [comment]
@akosyakov OK if it's because 2 events are sent.. I was not sure if it could be 2 sync called for some reason while it shouldn't be.

File: packages/monaco/src/browser/monaco-editor-model.ts (hunk 6 above)
Comment (hexa00) [reply packages/monaco/src/browser/monaco-editor-model.ts]: Yes seems like a good idea to me.

Sorry for the delay I missed that message.

File: packages/filesystem/src/browser/file-resource.ts (hunk 5 above)
Comment (akosyakov) [reply packages/filesystem/src/browser/file-resource.ts]: I've removed outdated check

File: packages/monaco/src/browser/monaco-editor-model.ts (hunk 6 above)
Comment (akosyakov) [reply packages/monaco/src/browser/monaco-editor-model.ts]: I've removed operations reverting, so if a user types something after applying operations then changes won't be lost.

**Comment by akosyakov** [comment]
I would like to merge it if nobody objects.

**Comment by hexa00** [comment]
@akosyakov didn't you want to remove the second commit ?

**Comment by akosyakov** [comment]
no, only `isOutdated` check, other checks are needed to fix https://github.com/theia-ide/theia/issues/1194

**Comment by hexa00** [comment]
Oops my bad it's not the one I Thought please merge

**Comment by akosyakov** [comment]
@hexa00 could you please approve 🙏
